### PR TITLE
drivers/samples: flash: shell: fix issues

### LINKED ISSUE #47576: undefined reference to `__device_dts_ord_20' When building with board hifive_unmatched on flash_shell samples
**Describe the bug**
I was trying to build the flash_shell samples using the board hifive_unmatched and the compilation failed with 

```
/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/drivers/flash/spi_nor.c:1230: 
undefined reference to `__device_dts_ord_20'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Board**
hifive_unmatched 

**What have you tried to diagnose or workaround this issue?**
- I verified the DT_DRV_COMPAT has named correctly
- I verified that the dt/bindings is in placed

**To Reproduce**
1. west init zephyrproject
2. cd zephyrproject
3. west update
4. cd zephyr
5. west build -b hifive_unmatched samples/drivers/flash_shell --pristine
6. see error

**Expected behavior**
Build success without error

**Impact**
(Showstoppers) This is blocking the PR https://github.com/zephyrproject-rtos/zephyr/pull/46443

**Logs and console output**
```
[09/07/2022 - 07:19.01] [boonkhai.boonkhai-container-55d96b56d8-fgzrj]
/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr
$ west build -b hifive_unmatched samples/drivers/flash_shell --pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr repository).
-- Application: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell
-- Found Python3: /build/zephyrproject/.venv/bin/python3.6 (found suitable exact version "3.6.15") found components: Interpreter 
-- Cache files will be written to: /nfs/site/home/boonkhai/.cache/zephyr
-- Zephyr version: 3.1.99 (/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr)
-- Found west (found suitable version "0.12.0", minimum required is "0.7.1")
-- Board: hifive_unmatched
-- Found host-tools: zephyr 0.13.2 (/build/zephyr-sdk-0.13.2)
-- Found dtc: /build/zephyr-sdk-0.13.2/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6")
-- Found toolchain: zephyr 0.13.2 (/build/zephyr-sdk-0.13.2)
-- Found BOARD.dts: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/boards/riscv/hifive_unmatched/hifive_unmatched.dts
-- Generated zephyr.dts: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/include/generated/devicetree_unfixed.h
-- Generated device_extern.h: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/include/generated/device_extern.h
-- Including generated dts.cmake file: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/dts.cmake
Parsing /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell/Kconfig
Loaded configuration '/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/boards/riscv/hifive_unmatched/hifive_unmatched_defconfig'
Merged configuration '/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell/prj.conf'
Configuration saved to '/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/.config'
Kconfig header saved to '/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 10.3.0
-- The CXX compiler identification is GNU 10.3.0
-- The ASM compiler identification is GNU
-- Found assembler: /build/zephyr-sdk-0.13.2/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc
-- Configuring done
-- Generating done
-- Build files have been written to: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build
-- west build: building application
[1/163] Preparing syscall dependency handling

[3/163] Generating include/generated/version.h
-- Zephyr version: 3.1.99 (/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr), build: zephyr-v3.1.0-1882-gb161554ee862
[49/163] Building C object CMakeFiles/app.dir/src/main.c.obj
/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c: In function 'do_write_unaligned':
/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c:304:23: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  304 |   off_t data_offset = (off_t)buf;
      |                       ^
/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c:317:7: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  317 |       (const void *)data_offset,
      |       ^
[153/163] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map 
: && ccache /build/zephyr-sdk-0.13.2/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc   zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/riscv/core/libarch__riscv__core.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/spi/libdrivers__spi.a  zephyr/drivers/flash/libdrivers__flash.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  zephyr/CMakeFiles/offsets.dir/./arch/riscv/core/offsets/offsets.c.obj  -L"/build/zephyr-sdk-0.13.2/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/rv64imac/lp64/medany"  -L/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -no-pie  -mcmodel=medany  -mabi=lp64  -march=rv64imac  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cd /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr && /usr/local/bin/cmake -E echo
/build/zephyr-sdk-0.13.2/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: zephyr/drivers/flash/libdrivers__flash.a(spi_nor.c.obj):/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/drivers/flash/spi_nor.c:1230: undefined reference to `__device_dts_ord_20'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/local/bin/cmake --build /nfs/png/disks/swuser_work_boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build

```

**Environment (please complete the following information):**
- Linux 
- Zephyr SDK
- zephyr-sdk-0.13.2

### PR DESCRIPTION
**Pull request by gmarull**
- The flash shell/sample shell included soc.h for no reason, remove it. Also removed a duplicate include.
- Fix hifive_unmatched issues with `jedec,spi-nor` (node could not be used)

Fixes #47576

minimal libc issues discovered here are being fixed in #47612

### CODE REVIEW

**Review by carlescufi — DISMISSED** [review]

**Review by carlocaione — DISMISSED** [review]

**Comment by gmarull** [comment]
Investigating current failure (also on `main`):

`west build -b hifive_unmatched samples/drivers/flash_shell -p`

```
/home/gmarull/ws/nordic/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c: In function 'do_write_unaligned':
/home/gmarull/ws/nordic/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c:304:23: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  304 |   off_t data_offset = (off_t)buf;
      |                       ^
/home/gmarull/ws/nordic/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c:317:7: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  317 |       (const void *)data_offset,
      |       ^
[153/163] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr/zephyr_pre0.map 
: && ccache /home/gmarull/zephyr-sdk/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc   zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/riscv/core/offsets/offsets.c.obj  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/riscv/core/libarch__riscv__core.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/spi/libdrivers__spi.a  zephyr/drivers/flash/libdrivers__flash.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/home/gmarull/zephyr-sdk/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/rv64imac/lp64/medany"  -L/home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -no-pie  -mcmodel=medany  -mabi=lp64  -march=rv64imac  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cd /home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr && /home/gmarull/.local/lib/python3.9/site-packages/cmake/data/bin/cmake -E echo
/home/gmarull/zephyr-sdk/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: zephyr/drivers/flash/libdrivers__flash.a(spi_nor.c.obj):/home/gmarull/ws/nordic/zephyrproject/zephyr/drivers/flash/spi_nor.c:1230: undefined reference to `__device_dts_ord_20'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by carlocaione** [comment]
@gmarull see #47576

**Comment by gmarull** [comment]
Note: there seems to be a bug in minimal lib w.r.t `off_t` definition, discussed with @stephanosio who will submit a fix.

File: lib/libc/minimal/include/sys/types.h [hunk 1]
```diff
@@ -33,25 +33,7 @@ typedef __SIZE_TYPE__ ssize_t;
...
-#elif defined(__XTENSA__)
-typedef int off_t;
-#elif defined(__sparc__)
-typedef int off_t;
-#elif defined(__mips)
-typedef int off_t;
-#else
-#error "The minimal libc library does not recognize the architecture!\n"
-#endif
+typedef long off_t;
```
Comment (carlescufi) [inline lib/libc/minimal/include/sys/types.h:36]: `typdef long int off_t`

**Review by carlescufi — DISMISSED** [review]

File: lib/libc/minimal/include/sys/types.h (hunk 1 above)
Comment (gmarull) [reply lib/libc/minimal/include/sys/types.h:36]: checkpatch doesn't like this (check previous runs)

**Review by carlocaione — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**